src/rbac/default_role_manager.rs

```diff
@@ -0,0 +1,353 @@
+        let mut name = name.to_owned();
+        if let Some(domain_val) = domain {
+            name = format!("{}::{}", domain_val, name);
+        }
+        if !self.has_role(&name) {
+            return vec![];
+        }
+        let role = self.create_role(&name);
+
+        if let Some(domain_val) = domain {
+            return role
+                .borrow()
+                .get_roles()
+                .iter()
+                .map(|x| x[domain_val.len() + 2..].to_string())
+                .collect();
+        } else {
+            return role.borrow().get_roles();
+        }
+    }
+
+    fn get_users(&self, name: &str, domain: Option<&str>) -> Vec<String> {
+        let mut name = name.to_owned();
+        if let Some(domain_val) = domain {
+            name = format!("{}::{}", domain_val, name);
+        }
+        if !self.has_role(&name) {
+            return vec![];
+        }
+
+        let mut names: Vec<String> = vec![];
+        for (_key, role) in self.all_roles.iter() {
+            if role.borrow().has_direct_role(&name) {
+                names.push(role.borrow().name.clone());
+            }
+        }
+        if let Some(domain_val) = domain {
+            return names
+                .iter()
+                .map(|x| x[domain_val.len() + 2..].to_string())
+                .collect();
+        }
+        return names;
+    }
+
+    fn print_roles(&self) {
+        println!("current role manager roles: {:?}", self.all_roles.clone());
+    }
+
+    fn clear(&mut self) {
+        self.all_roles = HashMap::new();
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Role {
+    name: String,
+    roles: Vec<Rc<RefCell<Role>>>,
+}
+
+impl Role {
+    fn new(name: String) -> Self {
+        return Role {
+            name,
+            roles: vec![],
+        };
+    }
+
+    fn add_role(&mut self, other_role: Rc<RefCell<Role>>) {
+        for role in self.roles.iter() {
+            if role.borrow().name == other_role.borrow().name {
+                return;
+            }
+        }
+        self.roles.push(other_role);
+    }
+
+    fn delete_role(&mut self, other_role: Rc<RefCell<Role>>) {
+        if let Some(pos) = self
+            .roles
+            .iter()
+            .position(|x| x.borrow().name == other_role.borrow().name)
+        {
+            self.roles.remove(pos);
+        }
+    }
+
+    fn has_role(&self, name: &str, hierarchy_level: usize) -> bool {
+        if self.name == name {
+            return true;
+        }
+        if hierarchy_level <= 0 {
+            return false;
+        }
+        for role in self.roles.iter() {
+            if role.borrow().has_role(name, hierarchy_level - 1) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    fn get_roles(&self) -> Vec<String> {
+        let mut names = vec![];
+        for role in self.roles.iter() {
+            names.push(role.borrow().name.clone());
+        }
```

Comment:
A more idiomatic way to do this is with map + collect

Reply:
Thanks, I recently watched Pascal Hertleif's talk about declarative programming in rust, it better to use iterator then a for loop, will fix later.